mlflow/pipelines/steps/train.py

```diff
@@ -807,22 +807,23 @@ def objective(X_train, y_train, validation_df, hyperparameter_args, on_worker=Fa
             early_stop_fn = getattr(importlib.import_module(train_module_name), early_stop_fn_name)
             fmin_kwargs["early_stop_fn"] = early_stop_fn
         best_hp_params = fmin(**fmin_kwargs)
-        best_hp_estimator_loss = hp_trials.best_trial["result"]["loss"]
-        hardcoded_estimator_loss = objective(
-            X_train, y_train, validation_df, estimator_hardcoded_params
-        )
         best_hp_params = space_eval(search_space, best_hp_params)
+        best_hp_estimator_loss = hp_trials.best_trial["result"]["loss"]
+        if len(estimator_hardcoded_params) > 1:
+            hardcoded_estimator_loss = objective(
+                X_train, y_train, validation_df, estimator_hardcoded_params
+            )
 
-        if best_hp_estimator_loss < hardcoded_estimator_loss:
-            best_hardcoded_params = {
-                param_name: param_value
-                for param_name, param_value in estimator_hardcoded_params.items()
-                if param_name not in best_hp_params
-            }
-        else:
-            best_hp_params = {}
-            best_hardcoded_params = estimator_hardcoded_params
-
+            if best_hp_estimator_loss < hardcoded_estimator_loss:
+                best_hardcoded_params = {
+                    param_name: param_value
+                    for param_name, param_value in estimator_hardcoded_params.items()
+                    if param_name not in best_hp_params
+                }
+            else:
+                best_hp_params = {}
+                best_hardcoded_params = estimator_hardcoded_params
+        best_hardcoded_params = {}
```

Comment:
Looks like we reset `best_hardcoded_params` after the if / else block, even though both branches of the block set `best_hardcoded_params`. Is this intended? If this is an error (seems likely), do we have test coverage to catch it?

Reply:
good catch - i think this would have broken `test_train_step_with_tuning_best_parameters`